### PR TITLE
[SYCL] Split 'double type' test into two tests, 1) <test>.cpp; 2) <test>_aspect-fp64.cpp for llvm-test-suite

### PR DESCRIPTION
**Pull request by myler**
'double' type are not native supported in Intel Gen11/12 and next Gens GPUs in the future, to disable particularly big number of important tests is not reasonable, so split 'double' part into two separate test file.

- split double tests in AtomicRef directory
- split double tests in Basic directory
- split double tests in DeprecatedFeatures directory
- split double tests in DeviceLib directory
- split double tests in ESIMD directory
- split double tests in GroupAlgorithm directory
- split double tests in InlineAsm directory
- split double tests in KernelParams directory
- split double tests in Regression directory
- split double tests in SpecConstants directory
- split double tests in SubGroup directory
- split double tests in USM directory
- add `REQUIRES:aspect-fp64` to restrict some 'double' type tests "_double.cpp", "_fp64.cpp", "_fp_extra.cpp", etc.

### CODE REVIEW

**Comment by myler** [comment]
Failures are unrelated:

[2022-08-17T07:35:47.081Z] Failed Tests (1):
[2022-08-17T07:35:47.081Z]   SYCL :: InvokeSimd/invoke_simd_conv.cpp

[2022-08-17T07:41:39.229Z] Unexpectedly Passed Tests (2):
[2022-08-17T07:41:39.229Z]   SYCL :: SubGroup/reduce_spirv13.cpp
[2022-08-17T07:41:39.229Z]   SYCL :: SubGroup/reduce_spirv13_fp64.cpp

File: SYCL/ESIMD/api/bin_and_cmp_ops_heavy_aspect-fp64.cpp
Comment (v-klochkov) [inline SYCL/ESIMD/api/bin_and_cmp_ops_heavy_aspect-fp64.cpp]: If aspect::fp64 is checked for 'double', then shouldn't 'aspect::fp16 be also checked for 'half' ?

File: SYCL/USM/copy.cpp
Comment (aelovikov-intel) [inline SYCL/USM/copy.cpp]: Why isn't it enough to make early returns from this function, without creating an extra .cpp file? Same for other places.

File: SYCL/ESIMD/api/bin_and_cmp_ops_heavy_aspect-fp64.cpp
Comment (myler) [reply SYCL/ESIMD/api/bin_and_cmp_ops_heavy_aspect-fp64.cpp]: We have many aspects in sycl spec 2020, but we don't plan to separate files for all of them. AFAIK, Intel newer GPUs will no longer support fp64 (double) natively, so that's why we need to split 'aspect::fp64'.

Tagging @pvchupin who suggests this change instead of disabling all important tests.

File: SYCL/USM/copy.cpp
Comment (myler) [reply SYCL/USM/copy.cpp]: Not sure you are asking for `dev_dst1 = (kind1 == alloc::device);`?
I guess you mean returns from this function if the device `!has(sycl::aspect::fp64)`?
The JIT compiler raises unsupported 'double' type error in compiling no matter we have `if/else` statement to return early. We have to split them into two TUs.

File: SYCL/ESIMD/api/bin_and_cmp_ops_heavy_aspect-fp64.cpp
Comment (v-klochkov) [reply SYCL/ESIMD/api/bin_and_cmp_ops_heavy_aspect-fp64.cpp]: I am Ok with not splitting/separating fp16 for ESIMD test. 
For more general and non-ESIMD tests (if there are such in this PR), using fp16 aspect check would make more sense, IMO. Thank you.

File: SYCL/USM/copy.cpp
Comment (aelovikov-intel) [reply SYCL/USM/copy.cpp]: How do customers solve this? Imagine I have fast double algorithm and slow float algorithm, and I want to do host-based dispatch to run one based on the current device. Are you saying this is impossible?

Comment (pvchupin) [reply SYCL/USM/copy.cpp]: Do we have compiler implementation lagging behind? Probably not everything from SYCL2020 in this regard is implemented yet.

Tagging @gmlueck

Comment (gmlueck) [reply SYCL/USM/copy.cpp]: Yes, this is a bug in DPC++, and we have had customer complaints about it.  Applications should not be required to put their "double" kernels in a separate source file.  The SYCL 2020 specification requires that this work even when there is a single source file.  The proposed solution is the design that is outlined in [OptionalDeviceFeatures](https://github.com/intel/llvm/blob/sycl/sycl/doc/design/OptionalDeviceFeatures.md).

Comment (pvchupin) [reply SYCL/USM/copy.cpp]: @AlexeySachkov do we plan to address it soon? Otherwise we should split to keep testing specific underlying features properly.

File: SYCL/ESIMD/api/bin_and_cmp_ops_heavy_aspect-fp64.cpp
Comment (myler) [reply SYCL/ESIMD/api/bin_and_cmp_ops_heavy_aspect-fp64.cpp]: Will do 'aspect::fp16 to check half' in another PR later.

Comment (v-klochkov) [reply SYCL/ESIMD/api/bin_and_cmp_ops_heavy_aspect-fp64.cpp]: Ok, Thank you!

**Review by v-klochkov — DISMISSED** [review]

File: SYCL/USM/copy.cpp
Comment (aelovikov-intel) [reply SYCL/USM/copy.cpp]: I still think splitting like this increases maintainability burden greatly. I'd rather see something like:

```
// test.cpp
// RUN: ... -DENABLE_FP64=false
constexpr bool EnableFP64 = ENABLE_FP64;

void test() {
   // non-fp64-case
   do_smth();

   if constexpr (EnableFP64)
     do_smth64();
}

// test-fp64.cpp
; RUN: ... -DENABLE_FP64=true
#include <test-fp.cpp>
```

Comment (myler) [reply SYCL/USM/copy.cpp]: @aelovikov-intel, that's good solution!
The preprocess step is able to strip out those fp64 code, and no complain is raised from GPU compiler `error: Double type is not supported on this platform.` any more.
I will update this change in 24 files `<test>_aspect-fp64.cpp` very soon.

Comment (aelovikov-intel) [reply SYCL/USM/copy.cpp]: Thanks!

**Comment by myler** [comment]
Unrelated failures:
```
[2022-08-23T09:33:09.407Z] Failed Tests (5):
[2022-08-23T09:33:09.407Z]   SYCL :: DiscardEvents/discard_events_accessors.cpp
[2022-08-23T09:33:09.407Z]   SYCL :: DiscardEvents/discard_events_using_assert.cpp
[2022-08-23T09:33:09.407Z]   SYCL :: DiscardEvents/discard_events_using_assert_ndebug.cpp
[2022-08-23T09:33:09.407Z]   SYCL :: DiscardEvents/discard_events_usm.cpp
[2022-08-23T09:33:09.407Z]   SYCL :: DiscardEvents/discard_events_usm_ooo_queue.cpp
```

File: SYCL/Basic/buffer/buffer.cpp
Comment (aelovikov-intel) [inline SYCL/Basic/buffer/buffer.cpp]: Maybe just me, but this is hard to read. Also, I think there is a functional change here. Consider the case when bool/int are ok and double is wrong. We won't even take the first if so the double check will not be done at all. I'd change like this:

```
  bool Passed = true;
  Passed &= (bool_vector[i] == true);
  Passed &= (int_vector[i] == 3);
  #ifdef ENABLE_FP64
  Passed &= (double_vector[i] == 7.5);
  #endif
```

File: SYCL/Basic/buffer/buffer.cpp
Comment (aelovikov-intel) [inline SYCL/Basic/buffer/buffer.cpp]: Unrelated change?

File: SYCL/ESIMD/api/bin_and_cmp_ops_heavy_aspect-fp64.cpp
Comment (aelovikov-intel) [inline SYCL/ESIMD/api/bin_and_cmp_ops_heavy_aspect-fp64.cpp]: I don't think we need to duplicate the title comments in fp64 versions. I'd change all of the them to a generic

```
Enable FP64 part of <orig_test_name.cpp>. To be removed once DPC++
supports optional device features and the code could be enabled
unconditionally without causing failures in speculative compilation
of the kernels.
```

File: SYCL/ESIMD/ext_math.cpp
Comment (aelovikov-intel) [inline SYCL/ESIMD/ext_math.cpp:479]: Why is this float guarded?

File: SYCL/ESIMD/regression/Inputs/dgetrf.hpp
Comment (aelovikov-intel) [inline SYCL/ESIMD/regression/Inputs/dgetrf.hpp]:
```suggestion
#ifdef ENABLE_FP64
using fptype = double;
#else
using fptype = float;
#endif
```

File: SYCL/ESIMD/regression/Inputs/dgetrf.hpp
Comment (aelovikov-intel) [inline SYCL/ESIMD/regression/Inputs/dgetrf.hpp]: @intel/dpcpp-esimd-reviewers , is this change (192->256) ok?

File: SYCL/ESIMD/regression/dgetrf_8x8.cpp
Comment (aelovikov-intel) [inline SYCL/ESIMD/regression/dgetrf_8x8.cpp]: Please change to use C++ style type aliases `using fptype = ....`.

File: SYCL/InlineAsm/asm_float_add.cpp
Comment (aelovikov-intel) [inline SYCL/InlineAsm/asm_float_add.cpp]: Why this change? Is there a second part of it somewhere in this patch?

File: SYCL/InlineAsm/asm_float_imm_arg.cpp
Comment (aelovikov-intel) [inline SYCL/InlineAsm/asm_float_imm_arg.cpp]: Likewise - why this change?

File: SYCL/KernelParams/union_kernel_param.cpp
Comment (aelovikov-intel) [inline SYCL/KernelParams/union_kernel_param.cpp]: Please use C++ style type alias.

File: SYCL/SpecConstants/2020/handler-api.cpp
Comment (aelovikov-intel) [inline SYCL/SpecConstants/2020/handler-api.cpp]: I don't think we really need to move it to after `custom_type_id` (imagine we'll remove `#ifdef` without moving it back), but I can live with it.

Same below.

File: SYCL/Basic/buffer/buffer.cpp
Comment (myler) [reply SYCL/Basic/buffer/buffer.cpp]: I just applied the auto generated patch by the clang-format-check testing.

File: SYCL/ESIMD/api/bin_and_cmp_ops_heavy_aspect-fp64.cpp
Comment (myler) [reply SYCL/ESIMD/api/bin_and_cmp_ops_heavy_aspect-fp64.cpp]: Applied this to all new added fp64 sources.

File: SYCL/ESIMD/regression/Inputs/dgetrf.hpp
Comment (myler) [reply SYCL/ESIMD/regression/Inputs/dgetrf.hpp]: Revert the change back to `192`, this is an internal requirement, should not merge to public repo.

File: SYCL/SpecConstants/2020/handler-api.cpp
Comment (myler) [reply SYCL/SpecConstants/2020/handler-api.cpp]: Moved `fp64` code block before `custom_type_id`

**Comment by myler** [comment]
Unrelated failure:
```
[2022-08-24T04:03:12.640Z] Failed Tests (1):
[2022-08-24T04:03:12.640Z]   SYCL :: DeprecatedFeatures/buffer_interop.cpp
```

File: SYCL/Basic/buffer/buffer.cpp
Comment (aelovikov-intel) [reply SYCL/Basic/buffer/buffer.cpp]: It looks like previously line 528 scope was closed at line 573 and it isn't closed there anymore. Hard to track that in the web-browser though, so I might be wrong.

File: SYCL/ESIMD/api/bin_and_cmp_ops_heavy_aspect-fp64.cpp
Comment (aelovikov-intel) [reply SYCL/ESIMD/api/bin_and_cmp_ops_heavy_aspect-fp64.cpp]: But what about this one on lines 21-30 ?

File: SYCL/ESIMD/api/saturation_smoke_aspect-fp64.cpp
Comment (aelovikov-intel) [inline SYCL/ESIMD/api/saturation_smoke_aspect-fp64.cpp]: This looks wrong.

File: SYCL/InlineAsm/asm_float_add_aspect-fp64.cpp
Comment (aelovikov-intel) [inline SYCL/InlineAsm/asm_float_add_aspect-fp64.cpp]: I don't think this is accurate. Removal of the ENABLE_FP64 would result in dropping one of the cases in the main file, not executing them both.

File: SYCL/ESIMD/api/saturation_smoke_aspect-fp64.cpp
Comment (myler) [reply SYCL/ESIMD/api/saturation_smoke_aspect-fp64.cpp]: This change is still required by the clang-format-check patch.

File: SYCL/ESIMD/ext_math.cpp
Comment (myler) [reply SYCL/ESIMD/ext_math.cpp:479]: Re-enable those IEEE-conformant single precision checks.

File: SYCL/ESIMD/api/saturation_smoke_aspect-fp64.cpp
Comment (aelovikov-intel) [reply SYCL/ESIMD/api/saturation_smoke_aspect-fp64.cpp]: Shorten the "---" parts and it will fit one line.

File: SYCL/Basic/buffer/buffer.cpp
Comment (aelovikov-intel) [reply SYCL/Basic/buffer/buffer.cpp]: I don't think this has been addressed...

File: SYCL/ESIMD/regression/dgetrf_8x8_aspect-fp64.cpp
Comment (aelovikov-intel) [inline SYCL/ESIMD/regression/dgetrf_8x8_aspect-fp64.cpp]: I think I've mentioned that before, maybe for another test. This is incorrect as doubles aren't filtered out in the included sources (to the best of my understanding). Instead, we have a selection between float/double and they cannot co-exist at the same time.

File: SYCL/InlineAsm/asm_float_imm_arg_aspect-fp64.cpp
Comment (aelovikov-intel) [inline SYCL/InlineAsm/asm_float_imm_arg_aspect-fp64.cpp]: Same as above.

File: SYCL/SubGroup/generic-shuffle_aspect-fp64.cpp
Comment (aelovikov-intel) [inline SYCL/SubGroup/generic-shuffle_aspect-fp64.cpp]: This looks wrong. Can you please fix such dangling formatting throughout the whole PR?

File: SYCL/ESIMD/regression/dgetrf_8x8_aspect-fp64.cpp
Comment (myler) [reply SYCL/ESIMD/regression/dgetrf_8x8_aspect-fp64.cpp]: Fixed in `dgetrf.cpp` by using function template.

File: SYCL/InlineAsm/asm_float_imm_arg_aspect-fp64.cpp
Comment (myler) [reply SYCL/InlineAsm/asm_float_imm_arg_aspect-fp64.cpp]: Fixed in `asm_float_imm_arg.cpp` by using function template.

File: SYCL/Basic/buffer/buffer.cpp
Comment (myler) [reply SYCL/Basic/buffer/buffer.cpp]: Sorry for the mistake, fixed the in-correct brackets `{`.

**Comment by myler** [comment]
Related failure:
```
[2022-08-30T07:31:19.596Z] Failed Tests (1):
[2022-08-30T07:31:19.596Z]   SYCL :: ESIMD/aot_mixed.cpp
```
This error `LLVM ERROR: vc::translateBuild: unsupported platform` cannot be reproduced.
```
Running on Intel(R) Iris(R) Xe Graphics
Running ESIMD kernel...
TEST Passed
```


Unrelated failures:
```
[2022-08-30T08:08:17.655Z] Failed Tests (4):
[2022-08-30T08:08:17.655Z]   SYCL :: ESIMD/api/replicate_smoke.cpp
[2022-08-30T08:08:17.655Z]   SYCL :: ESIMD/api/simd_view_select_2d_int.cpp
[2022-08-30T08:08:17.655Z]   SYCL :: ESIMD/ext_math.cpp
[2022-08-30T08:08:17.655Z]   SYCL :: ESIMD/matrix_transpose_glb.cpp
```

**Review by aelovikov-intel — DISMISSED** [review]

**Review by v-klochkov — DISMISSED** [review]
Ok for ESIMD. Thank you.

I also have a non-blocking comment regarding the naming.
IMO, it is better to stick to underscores and not mix them with '-'.
E.g. "functions_select_2d_core_aspect_fp64.cpp" looks much better to me than "functions_select_2d_core_aspect-fp64.cpp"

**Review by AlexeySachkov — DISMISSED** [review]
LGTM for spec constants

**Comment by myler** [comment]
> Ok for ESIMD. Thank you.
> 
> I also have a non-blocking comment regarding the naming. IMO, it is better to stick to underscores and not mix them with '-'. E.g. "functions_select_2d_core_aspect_fp64.cpp" looks much better to me than "functions_select_2d_core_aspect-fp64.cpp"

Renamed these new added file with only underscores.

**Comment by myler** [comment]
Hi @Pennycook,
Seems other subdirectories are already approved, but may I ask you to review changes in `SubGroup` and `GroupAlgorithm`, seems you are the code owner of them.

**Comment by myler** [comment]
Related failure:
```
[2022-08-31T02:01:02.817Z] Failed Tests (1):
[2022-08-31T02:01:02.817Z]   SYCL :: ESIMD/aot_mixed.cpp
```
This error `LLVM ERROR: vc::translateBuild: unsupported platform` cannot be reproduced.
```
Running on Intel(R) Iris(R) Xe Graphics
Running ESIMD kernel...
TEST Passed
```
Unrelated failures:
```
[2022-08-31T02:34:16.124Z] Failed Tests (4):
[2022-08-31T02:34:16.124Z]   SYCL :: ESIMD/api/replicate_smoke.cpp
[2022-08-31T02:34:16.124Z]   SYCL :: ESIMD/api/simd_view_select_2d_int.cpp
[2022-08-31T02:34:16.124Z]   SYCL :: ESIMD/ext_math.cpp
[2022-08-31T02:34:16.124Z]   SYCL :: ESIMD/matrix_transpose_glb.cpp
```

**Review by v-klochkov — DISMISSED** [review]

**Review by AlexeySachkov — DISMISSED** [review]

**Comment by romanovvlad** [comment]
A possible alternative solution is here: https://github.com/intel/llvm-test-suite/pull/1190/files

**Comment by myler** [comment]
Failures are known and unrelated.
Hi @v-klochkov 
According to the https://github.com/intel/llvm-test-suite/blob/intel/CONTRIBUTING.md#merge, need maintainer to merge the PR. Could you help me on this?

**Comment by v-klochkov** [comment]
> Failures are known and unrelated. Hi @v-klochkov According to the https://github.com/intel/llvm-test-suite/blob/intel/CONTRIBUTING.md#merge, need maintainer to merge the PR. Could you help me on this?

I did not merge this yet because of the comment from @romanovvlad (https://github.com/intel/llvm-test-suite/pull/1150#issuecomment-1232874469) and did not have clear understanding if the discussion here https://github.com/intel/llvm-test-suite/pull/1190 has got some solution/agreement.

That alternative approach used in https://github.com/intel/llvm-test-suite/pull/1190 has sympathy/votes from at least 3 people. 
And I would add my vote to there as well because device.has is specially designed and convenient way to run-or-skip hw specific features.

I gave my approval for this PR, but I'd rather let @romanovvlad to merge this patch if Vlad's comment was non-blocking

**Review by v-klochkov — APPROVED** [review]

**Comment by AlexeySachkov** [comment]
Personally, I would vote for device code split approach, because it allows us to avoid duplicating coverage for non-optional types. I can help with creating PRs similar to one that @romanovvlad published (the first one to go is #1241)

**Comment by pvchupin** [comment]
@v-klochkov, @myler , @romanovvlad , @AlexeySachkov 
I assume we decided to proceed with alternative approach partially done at #1241 , #1190?

@aelovikov-intel, can you please take a look at remaining ones on Windows? We currently have ~30 fails and it's really unsafe to merge anything else until these fixed.

**Comment by pvchupin** [comment]
@myler, should we close this PR in favor of alternatives in progress?

**Comment by myler** [comment]
> @myler, should we close this PR in favor of alternatives in progress?

Hi @pvchupin 

This solution tends to change the testing framework `REQURIES: aspect-fp64` and macros `#ifdef ENABLE_FP64`, not depends on SYCL standard `aspect::fp64` and feature `-fsycl-device-code-split=per_kernel`

#1190 tends to use SYCL style solution to handle `double` type tests.

If the final decision is to use `if (device.has(aspect::fp64))` + `-fsycl-device-code-split=per_kernel` to handle `double` type tests, I agree to close this one. But #1190 did not fix all tests, we may have at least 26 tests contain `double` type code.

**Comment by gmlueck** [comment]
> If the final decision is to use if (device.has(aspect::fp64)) + -fsycl-device-code-split=per_kernel to handle double type tests, I agree to close this one. But https://github.com/intel/llvm-test-suite/pull/1190 did not fix all tests, we may have at least 26 tests contain double type code.

That seems like a better solution to me.  Once we implement the [OptionalDeviceFeatures](https://github.com/intel/llvm/blob/sycl/sycl/doc/design/OptionalDeviceFeatures.md) design, we can remove the "-fsycl-device-code-split=per_kernel", which should be relatively easy.

Are you saying  there are 26 test failures that will not be fixed with this technique?  Do we know what the cause of those failures is?

**Comment by myler** [comment]
Close this PR as we have another alternative.

**Pull request closed without merging**
